buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2028 @@
+     * We set the max composite buffer capacity to the same, since it would otherwise be impossible to create a
+     * non-composite copy of the buffer.
+     */
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE - 8;
+    private static final Drop<CompositeBuffer> COMPOSITE_DROP = new Drop<>() {
+        @Override
+        public void drop(CompositeBuffer buf) {
+            for (Buffer b : buf.bufs) {
+                b.close();
+            }
+            buf.makeInaccessible();
+        }
+
+        @Override
+        public String toString() {
+            return "COMPOSITE_DROP";
+        }
+    };
+    private static final Buffer[] EMPTY_BUFFER_ARRAY = new Buffer[0];
+
+    private final BufferAllocator allocator;
+    private final TornBufferAccessors tornBufAccessors;
+    private Buffer[] bufs;
+    private int[] offsets; // The offset, for the composite buffer, where each constituent buffer starts.
+    private int capacity;
+    private int roff;
+    private int woff;
+    private int subOffset; // The next offset *within* a consituent buffer to read from or write to.
+    private ByteOrder order;
+    private boolean closed;
+    private boolean readOnly;
+
+    /**
+     * Compose the given sequence of sends of buffers and present them as a single buffer.
+     * <p>
+     * When a composite buffer is closed, all of its constituent component buffers are closed as well.
+     * <p>
+     * See the class documentation for more information on what is required of the given buffers for composition to be
+     * allowed.
+     *
+     * @param allocator The allocator for the composite buffer. This allocator will be used e.g. to service
+     * {@link #ensureWritable(int)} calls.
+     * @param sends The sent buffers to compose into a single buffer view.
+     * @return A buffer composed of, and backed by, the given buffers.
+     * @throws IllegalArgumentException if the given buffers have an inconsistent
+     * {@linkplain Buffer#order() byte order}.
+     * @throws IllegalStateException if one of the sends have already been received. The remaining buffers and sends
+     * will be closed and discarded, respectively.
+     */
+    @SafeVarargs
+    public static CompositeBuffer compose(BufferAllocator allocator, Send<Buffer>... sends) {
+        Buffer[] bufs = new Buffer[sends.length];
+        IllegalStateException ise = null;
+        for (int i = 0; i < sends.length; i++) {
+            if (ise != null) {
+                try {
+                    sends[i].close();
+                } catch (Exception closeExc) {
+                    ise.addSuppressed(closeExc);
+                }
+            } else {
+                try {
+                    bufs[i] = sends[i].receive();
+                } catch (IllegalStateException e) {
+                    ise = e;
+                    for (int j = 0; j < i; j++) {
```

Comment:
Does this need to be `j <= i` ?  In the case of an exception, the i'th element is the only one that doesn't get close() called on the buf or the send.

Reply:
No, `bufs[i]` will be `null` if `sends[i].receive()` throws.